Remove type subclassing

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24284 Merge ProfiledTensorType and TensorType
* **#24257 Remove type subclassing**
* #24169 remove CompleteTensorType

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

Differential Revision: [D16794035](https://our.internmc.facebook.com/intern/diff/D16794035)

zdevito committed:
Remove type subclassing

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

zdevito committed:
Update on "Remove type subclassing"

Remove type subclassing

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

gh-metadata: pytorch pytorch 24257 gh/zdevito/84/head

Krovatkin commented on test/test_jit.py:
```diff
@@ -8208,6 +8208,7 @@ def tensor_unifying(x, y, z):
         c = torch.zeros(2, 4, dtype=torch.float)
 
         graph = _propagate_shapes(tensor_unifying.graph, (a, b, c), False)
+        print(graph)
```
a stray print?

Krovatkin commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -614,19 +542,32 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
     auto srs = strides().merge(other->strides());
     auto gr = merge_primitive(requiresGrad(), other->requiresGrad());
-    return ProfiledTensorType::create(scalar_type, dev, sz, srs, gr);
+    auto zero = merge_primitive(autogradZero(), other->autogradZero());
```
does switching from an undefined to defined tensor make sense? Maybe, it should be an assert?
`TORCH_INTERNAL_ASSERT(autogradZero() == other->autogradZero())`
I actually went the same path and added the same field in my branch, but I wasn't sure if it makes sense for a tensor to switch.

zdevito replied:
In this case it is merging a tensor which is known to be zero with one that that may/may not be zero. So the result is a tensor which may/may not be zero, which is what this implements.

Krovatkin commented on aten/src/ATen/core/jit_type.h:
I wish we had this comment before, took me a while to figure out how all these pieces fit together 😢

Krovatkin commented on torch/csrc/jit/graph_executor_impl.h:
```diff
@@ -87,7 +87,7 @@ struct GraphExecutorImplBase {
     for (size_t i = 0; i < input_values.size(); ++i) {
       // propagate tensor types
-      if (input_values.at(i)->type()->cast<TensorType>()) {
+      if (input_values.at(i)->type()->isSubtypeOf(TensorType::get())) {
```
what's wrong with cast to `TensorType`?

Krovatkin reviewed (approved):
:shipit:

zdevito committed:
Update on "Remove type subclassing"

Remove type subclassing

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

gh-metadata: pytorch pytorch 24257 gh/zdevito/84/head

zdevito committed:
Update on "Remove type subclassing"

Remove type subclassing

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

gh-metadata: pytorch pytorch 24257 gh/zdevito/84/head

zdevito committed:
Update on "Remove type subclassing"

Remove type subclassing

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

gh-metadata: pytorch pytorch 24257 gh/zdevito/84/head
